### ISSUE TITLE
Unknown feature names are not reported if there are other errors

### ISSUE THREAD

**Issue by jminer** [rust-lang/rust#58390]
I just added an impl that I thought should be valid now with https://github.com/rust-lang/rfcs/pull/2451, but I got a compiler error. So I found the RFC and copied the feature name `re_rebalancing_coherence` from it. However, I still got the compile error, so I spent a while trying to figure out if the impl was really valid. Eventually, I tried commenting out the impl, and then (thanks to #52644), the compiler told me the feature was unknown (it has been renamed to `re_rebalance_coherence`). The impl was valid using the correct feature name. I didn't get the error saying the feature name was invalid because of the error caused by the feature name being invalid.

I think the compiler should report unknown feature names even if there are other errors. For example: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=e46fde7bf4076220df3005d702df9cbb doesn't report the feature name is invalid. With box syntax, it suggests "help: add #![feature(box_syntax)] to the crate attributes to enable" but the coherence error didn't.

Feel free to close this issue if you don't think it is worth fixing.

**Comment by euclio** [comment]
I'd like to work on this.

**Comment by euclio** [comment]
The challenge here is that the current logic for reporting unknown features doesn't get run until stability checking. So, if there's an error before then, then the check doesn't run. While it's possible for us to get language features up-front for error reporting, library features are collected much later. So, before we can report unknown feature names earlier, we have to figure out how to collect the library names earlier.

**Comment by estebank** [comment]
Current output:
```
error: `box_syntax` has been removed
 --> src/main.rs:5:13
  |
5 |     let x = box 5;
  |             ^^^^^
  |
help: use `Box::new()` instead
  |
5 -     let x = box 5;
5 +     let x = Box::new(5);
  |

error[E0635]: unknown feature `box_syntax2`
 --> src/main.rs:2:12
  |
2 | #![feature(box_syntax2)]
  |            ^^^^^^^^^^^
```
Tried [other](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c05b6e53ec04d2dd1f3e60479c16ce9d) [cases](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=9da687f952dc8950a18ac90914f143b0) to check that the error is still being emitted even if there are other errors.
